File: tests/test_structure.py
"""Tests for Phase 1: Structure."""

from __future__ import annotations

import os

from mycelium.config import AnalysisConfig
from mycelium.graph.knowledge_graph import KnowledgeGraph
from mycelium.phases.structure import run_structure_phase

FIXTURES_DIR = os.path.join(os.path.dirname(__file__), "fixtures")


class TestStructurePhase:
    def test_finds_csharp_files(self):
        kg = KnowledgeGraph()
        config = AnalysisConfig(repo_path=os.path.join(FIXTURES_DIR, "csharp_simple"))
        run_structure_phase(config, kg)

        files = kg.get_files()
        file_paths = {f["path"] for f in files}

        assert "AbsenceController.cs" in file_paths
        assert "AbsenceService.cs" in file_paths
        assert "IAbsenceService.cs" in file_paths
        assert "AbsenceModel.cs" in file_paths
        assert "IAbsenceRepository.cs" in file_paths
        assert "AbsenceRepository.cs" in file_paths
        assert "LeaveRequestValidator.cs" in file_paths
        assert "AbsenceException.cs" in file_paths
        assert len(files) == 8

    def test_detects_language(self):
        kg = KnowledgeGraph()
        config = AnalysisConfig(repo_path=os.path.join(FIXTURES_DIR, "csharp_simple"))
        run_structure_phase(config, kg)

        files = kg.get_files()
        for f in files:
            assert f["language"] == "cs"

    def test_counts_lines(self):
        kg = KnowledgeGraph()
        config = AnalysisConfig(repo_path=os.path.join(FIXTURES_DIR, "csharp_simple"))
        run_structure_phase(config, kg)

        files = {f["path"]: f for f in kg.get_files()}
        # AbsenceController.cs has 27 lines
        assert files["AbsenceController.cs"]["lines"] > 0

    def test_creates_folders(self):
        kg = KnowledgeGraph()
        config = AnalysisConfig(repo_path=os.path.join(FIXTURES_DIR, "csharp_simple"))
        run_structure_phase(config, kg)

        folders = kg.get_folders()
        assert len(folders) >= 1  # At least the root folder

    def test_skips_ignored_directories(self):
        """Ensure bin, obj, etc. are skipped."""
        kg = KnowledgeGraph()
        config = AnalysisConfig(repo_path=os.path.join(FIXTURES_DIR, "csharp_simple"))
        run_structure_phase(config, kg)

        file_paths = {f["path"] for f in kg.get_files()}
        for path in file_paths:
            assert "bin/" not in path
            assert "obj/" not in path
            assert ".git/" not in path

    def test_mixed_dotnet_finds_all_file_types(self):
        kg = KnowledgeGraph()
        config = AnalysisConfig(repo_path=os.path.join(FIXTURES_DIR, "mixed_dotnet"))
        run_structure_phase(config, kg)

        files = kg.get_files()
        file_paths = {f["path"] for f in files}

        # Should find .cs, .vb, .csproj, .vbproj, .sln files
        assert any(p.endswith(".cs") for p in file_paths)
        assert any(p.endswith(".vb") for p in file_paths)
        assert any(p.endswith(".csproj") for p in file_paths)
        assert any(p.endswith(".vbproj") for p in file_paths)
        assert any(p.endswith(".sln") for p in file_paths)

    def test_records_file_size(self):
        kg = KnowledgeGraph()
        config = AnalysisConfig(repo_path=os.path.join(FIXTURES_DIR, "csharp_simple"))
        run_structure_phase(config, kg)

        files = kg.get_files()
        for f in files:
            assert f["size"] > 0

    def test_vbnet_files_detected(self):
        kg = KnowledgeGraph()
        config = AnalysisConfig(repo_path=os.path.join(FIXTURES_DIR, "vbnet_simple"))
        run_structure_phase(config, kg)

        files = kg.get_files()
        vb_files = [f for f in files if f["language"] == "vb"]
        assert len(vb_files) == 5

File: crates/mycelium-core/tests/test_structure.rs
//! Phase 1: Structure phase integration tests.

mod common;

use common::*;

#[test]
fn discovers_csharp_files() {
    let r = run_structure("csharp_simple");
    let files = file_paths(&r.kg);
    assert!(
        files.iter().any(|f| f.ends_with(".cs")),
        "Should discover .cs files"
    );
    assert!(files.len() >= 8, "csharp_simple has 8 .cs files");
}

#[test]
fn detects_languages() {
    let r = run_structure("csharp_simple");
    let langs = file_languages(&r.kg);
    assert!(langs.contains("C#"), "Should detect C# language");
}

#[test]
fn counts_lines() {
    let r = run_structure("csharp_simple");
    for file_data in r.kg.get_files() {
        if let mycelium_core::graph::knowledge_graph::NodeData::File {
            language: Some(_),
            lines,
            ..
        } = file_data
        {
            assert!(*lines > 0, "Source files should have line counts > 0");
        }
    }
}

#[test]
fn creates_folders() {
    let r = run_structure("python_package");
    let folders = folder_paths(&r.kg);
    assert!(!folders.is_empty(), "Should create folder nodes");
}

#[test]
fn ignores_default_excluded_dirs() {
    let r = run_structure("python_simple");
    let files = file_paths(&r.kg);
    assert!(
        files.iter().all(|f| !f.contains("__pycache__")),
        "Should skip __pycache__"
    );
}

#[test]
fn multi_language_detection() {
    let r = run_structure("mixed_dotnet");
    let langs = file_languages(&r.kg);
    assert!(langs.contains("C#"), "Should detect C#");
    // .vb files won't have language if VB.NET analyser is unavailable
    // .sln and .csproj aren't source languages
}

#[test]
fn file_size_tracked() {
    let r = run_structure("csharp_simple");
    for file_data in r.kg.get_files() {
        if let mycelium_core::graph::knowledge_graph::NodeData::File { size, .. } = file_data {
            assert!(*size > 0, "Files should have non-zero size");
        }
    }
}

#[test]
fn language_filter() {
    let path = fixture_path("python_package");
    let config = mycelium_core::config::AnalysisConfig {
        repo_path: path.to_string_lossy().to_string(),
        languages: Some(vec!["Python".to_string()]),
        ..Default::default()
    };
    let mut kg = mycelium_core::graph::knowledge_graph::KnowledgeGraph::new();
    mycelium_core::phases::structure::run_structure_phase(&config, &mut kg);

    let langs = file_languages(&kg);
    assert!(langs.contains("Python"));
    // Should only contain Python files
    for file_data in kg.get_files() {
        if let mycelium_core::graph::knowledge_graph::NodeData::File {
            language: Some(lang),
            ..
        } = file_data
        {
            assert_eq!(lang, "Python", "Filter should only include Python");
        }
    }
}
